### ISSUE TITLE
SINGLE PASS STEREO RENDERING on mobile VR is STILL not functioning properly with Unity 2017.2 or 2017.3

### ISSUE THREAD

**Issue by 01GOD** [Unity-Technologies/PostProcessing#497]
Post processing stack 2 is STILL doing the same COMPLETELY UNUSABLE single view render when using single pass stereo rendering on Android mobile VR, when rendering for Gear VR. 

Basically one viewport renders gray and the other viewport has no effects. Imagine it is simply some line that needs to be added to the PP stack to enable single pass stereo support. Although may be something in the Unity rendering engine. 

Unity version is 2017.3 and 2017.3.1 on Windows 10. Seems to be a well known problem among most content creators trying to do single pass stereo rendering using ANY post effects with this or any asset on the asset store. 

NOTE: Chman should NOT close this UNTIL the post processing stack ACTUALLY functions properly with single pass stereo rendering on mobile VR. Preferably at least in the most recent release of 2017.3.1. 
ALSO don't merge this with a previous issue notification, UNLESS ABSOLUTELY SURE that the previous issue notification was NOT closed already and is STILL VERY ACTIVELY AND EARNESTLY BEING WORKED ON TODAY. 

Thanks in advance!

**Comment by Chman** [comment]
> Imagine it is simply some line that needs to be added to the shaders in the PP stack to enable single pass stereo support.

We accept PRs.

> NOTE: Chman should NOT close this UNTIL the post processing stack ACTUALLY functions properly with single pass stereo rendering on mobile VR. Preferably at least in the most recent release of 2017.3.1.
ALSO don't merge this with a previous issue notification, UNLESS ABSOLUTELY SURE that the previous issue notification was NOT closed already and is STILL VERY ACTIVELY AND EARNESTLY BEING WORKED ON TODAY.

Please stop writing in all caps. Being agressive doesn't make people want to help you out. And thanks for the personal attack. You're referring to #473 which was closed because as clearly stated in the issue it was a duplicate of #468 which is still opened (and besides, these two were reported for desktop VR, not mobile VR). We don't close valid issues, we close duplicated & fixed issues. If an issue is still opened it means that someone will get to it eventually.

**Comment by robbiesri** [comment]
I'll take the heat for this, no need to bang on Chman (or anyone else really).  I haven't had time to work on this.  I should be opening up some time in the coming weeks, so I'll be able to take a look.

>  Imagine it is simply some line that needs to be added to the shaders in the PP stack to enable single pass stereo support.

Nope.  There are a few systems to re-work.  RenderTextures need to be properly allocated.  Screen-space textures need to be re-declared.  Screen-space samples need to be re-configured.  Interpolants need to account for stereo eye propogation. Blits/copies need to properly account for texture arrays. Each effect has to be tested.

**Comment by nhaskins** [comment]
Good to know this issue is on the radar. I was using multi pass to get bloom for GearVR on 2017.3 but it's pretty expensive.  Looking forward to progress on the single pass post FX front. 

Thanks!

**Comment by nshelton** [comment]
Excellent I'm really excited for this. I was able get Bloom working with single pass but there are some artifacts. I'm working on getting TheWaveVR into single-pass stereo but this is a blocker since we love the postprocessing stack. 

I found a couple of things that might help. It seems mainly to come from the mip pyramid inside
`UnityEngine.Rendering.PostProcessing.Bloom.Render(PostProcessRenderContext context)`

1) the mip pyramid buffers are half the width they should be. For some reason `context.screenWidth` is the width of an eye, but we have a two eye buffer. So It looks a lot better starting with 2x width.

2) The upsample is smearing to the right, and it's worse in the right eye. It seems to be fine on the downsample. 
![smear](https://user-images.githubusercontent.com/441629/37180969-a65c0ac8-22f0-11e8-9c69-1405df4a5c3a.PNG)
Spent some time looking at the filters and messing with their parameters but everything seemed legit. 
it does seem a bit strange that `Sampling.hlsl` is using `UnityStereoTransformScreenSpaceTex` to perform these filters, since in theory downsampling/upsampling could treat the two eyes as just a normal buffer. but I guess that prevents smearing across the eyes. But this could be doing unneeded uv transforms.

(using 2017.3.0f3 with  821f007 )

**Comment by 01GOD** [comment]
Seriously hope that Unity is NOT starting to outsource to india. Come on guys. Quality would nosedive and would cost more in hassles than the company would foolishly believe it's saving.

**Comment by 01GOD** [comment]
Recommend getting Keijiro at Unity Tokyo participating on this project if he's not involved already. Seems to be in his historical scope of skillset. Would likely be easy for him.

**Comment by 01GOD** [comment]
@nshelton Do a search in the repo for "single". Last night I modded a file that had a comment about doubling the width of the render texture, although haven't tested it yet. Also worth noting that using IL2CPP seems to clear some errors.

**Comment by 01GOD** [comment]
@nshelton Found the file. PostProcessing/Runtime/PostProcessRenderContext.cs

Have a look. Hopefully can get a fully functional workaround going while waiting for version 2 to come out of beta.

**Comment by robbiesri** [comment]
> Last night I modded a file that had a comment about doubling the width of the render texture, although haven't tested it yet

I wrote that.  Doubling the texture width is a last resort, because on some platforms, the double-wide RenderTexture is not double the width of a single eye (because of POT mip requirements on some platforms). On newer version of Unity, we prefer to use a descriptor provided by the engine, detailing the stereo texture setup.  And that descriptor works across multiple stereo rendering modes, instead of doubling, which only works with double-wide.

tl;dr - Don't do that.

**Comment by robbiesri** [comment]
@nshelton Thanks for investigating that.  I think I've actually noticed some weird Bloom issues too, but I never was able to pin down a good repro.  I think Issue #388 is another reference point.

From taking a quick look at the code, I think I see what happens.  I originally had the code allocating the _texture_ based on the stereo size, but then the shader parameters were looking at the logical screen size.  At some point, it looks like the code changed, and the texture size was dropped.  Not surprising, I wasn't explicitly clear in the code that the texture size needed to be preserved. I'll get a PR ready to fix it.

**Comment by robbiesri** [comment]
> Seriously hope that Unity is NOT starting to outsource to india.
Nephew, what is you doing?

**Comment by robbiesri** [comment]
> Seriously hope that Unity is NOT starting to outsource to india.

Well, that's a new one. 😆 🤣

**Comment by 01GOD** [comment]
@robbiesri So if a person knows the multiplier for the render texture width, can simply multiply the texture width in PostProcessing/Runtime/PostProcessRenderContext.cs by that number and it will definitely build properly and the effects will be visible? Both viewports will render properly?

**Comment by 01GOD** [comment]
@nshelton "I was able get Bloom working with single pass but there are some artifacts. " 

Was that on mobile? If so, how? Because when it does build, it has been rendering only one viewport and that viewport didn't have any effects.

**Comment by 01GOD** [comment]
Did a test of doubling the width in the specified area of PostProcessing/Runtime/PostProcessRenderContext.cs and that didn't do anything. Seems that there is possibly an error in the Unity version checking system of the PPS. Doesn't seem to call anything special based on 2017.2 and above. 
To be concise about it, this code seems to be non-functional and personally, "don't have a robust way of determining" sounds like a possible excuse to not make a robust way: 

"
// Single-pass is only supported with 2017.2+ because
                    // that is when XRSettings.eyeTextureDesc is available.
                    // Without it, we don't have a robust method of determining
                    // if we are in single-pass.  Users can just double the width
                    // here if they KNOW they are using single-pass.

		    width = XRSettings.eyeTextureWidth;
                    height = XRSettings.eyeTextureHeight;
"

**Comment by 01GOD** [comment]
I'm reminded daily of what a horrible mistake it is for companies to outsource ANY software engineering to india, actually. Error after error after error. Crash after crash. At least it's not as bad as outsourcing ux to india...that is more hilariously bad...at least that is easier to point at and laugh though.

**Comment by Anton-Tril** [comment]
Hello there! Am I doing something wrong or TXAA does not work in SinglePass for Windows builds too?

**Comment by robbiesri** [comment]
Hey @Anton-Tril I opened #502 for ya, instead of muddying one thread.

**Comment by 01GOD** [comment]
Sigh...case in point...windows...outsourcing to india ruined that.

**Comment by andybak** [comment]
This looks promising but the trail has gone cold: https://github.com/Unity-Technologies/PostProcessing/commit/6268df318987e7ef3e434a70d4a0987911c5f248

**Comment by soraryu** [comment]
Experiencing the same issue on Oculus Go with Unity 2018.2.11f1 - left eye is grey, right eye no PostFX.

**Comment by fabiopolimeni** [comment]
Guys any news on this? If the Lightweight rendering pipeline is supposed to be for mobile and VR, having the post process effects broken doesn't really help the adoption.

Any ETA on this? (my god I sound like a producer now :P)
Thanks for the great work anyway ;)

**Comment by AlexandreFiset** [comment]
Here for an ETA as well :) (Oculus Go still broken)

**Comment by jonny2027** [comment]
Would love to know if this will be fixed soon, the issue is present on Oculus Quest as well.

**Comment by OptrixAU** [comment]
Still having issues on Oculus Quest - when you enable post-processing, one eye goes completely blank (the standard response on Oculus to having one or more shaders that aren't 100% compliant with stereo rendering). 

Note though that I find single-pass rendering rather touchy on Quest in all scenarios - if I turn on Deferred Rendering in Unity, it has a similar problems even without the post-processing stack, so this bug may lie in Unity.

Incredibly frustrating, since I've got an application that really benefits from the application of bloom, but multi-pass rendering is uncomfortably slow. I currently have to disable post-processing on Android builds, which provides a significantly poorer visual experience.

**Comment by viknesh2020** [comment]
@OptrixAU  Facing the same issue. PPS is still broken on Quest.

**Comment by OptrixAU** [comment]
I've noticed that having the PPS on doesn't cause major issues if you're using the LWRP (Lightweight Rendering Pipline) in Unity 2019. The issue with one eye not rendering seems to be resolved.

However, I still don't appear to be able to get bloom to actually **work** on the device. It takes approximately 20 years to compile the Uber shader (the first time at least, thank goodness it's a bit faster after that), but I don't see any bloom, even after ensuring that HDR rendering is enabled on the device.

So with LWRP you can at least get a **usable** build - you just don't get any benefit from the post-processing. There also doesn't appear to be performance impact, so I'm guessing it's actually been disabled on Quest. I've heard some rumors around that it's because the Quest is returning something incorrect when PPS is checking for texture support - worth looking into, but I haven't had time as yet.
